tests/tests.py

```diff
@@ -136,6 +137,179 @@ def test_application_update(self):
         cmd = "kubectl delete -f ../examples/multitenancy/hello-world/hello-world-service-composition-localchart.yaml --kubeconfig=../kubeplus-saas-provider.json"
         out, err = TestKubePlus.run_command(cmd)
 
+    def test_application_upgrade(self):
+        # imports
+        from kubernetes import config
+        from kubernetes.client import Configuration
+        from kubernetes.client.api import core_v1_api
+        from kubernetes.stream import portforward
+        import select
+        from bs4 import BeautifulSoup
+
+        # assume appropriate plugins installation and PATH update
+
+        # helper methods
+
+        # from kubernetes python client example -- pod_portforward.py
+        def make_http_request(port):
+            http = pf.socket(port)
+            http.setblocking(True)
+            http.sendall(b'GET /users HTTP/1.1\r\n')
+            http.sendall(b'Host: 127.0.0.1\r\n')
+            http.sendall(b'Connection: close\r\n')
+            http.sendall(b'Accept: */*\r\n')
+            http.sendall(b'\r\n')
+            response = b''
+            while True:
+                select.select([http], [], [])
+                data = http.recv(1024)
+                if not data:
+                    break
+                response += data
+            http.close()
+            
+            error = pf.error(port)
+            if error is None:
+                print("No port forward errors on port %d." % port)
+            else:
+                print(f"Port {port} has the following error: {error}")
+            http.close()
+            return response.decode('utf-8')
+        
+        # get number of users from HTML response
+        def count_users(response):
+            num_users = 0
+            if "No users" not in response:
+                soup = BeautifulSoup(response, 'lxml')
```

Comment:
Do we need to include any library in requirements.txt for BeautifulSoup?

Reply:
Yes, I added bs4 (and the dependent lxml module) and kubernetes to the requirements.txt just now.